File: go-impl/handlers/comic.go
package handlers

import (
	"fmt"
	"net/http"
	"strings"
	"teki/collection/database"
	"teki/collection/models"

	"github.com/gofiber/fiber/v2"
	"github.com/golang-jwt/jwt/v4"
)

type ErrorResponse struct {
	Message string `json:"message" example:"An error occurred"`
}

// GetAllComics is a function to get all comics data from database
//
// @Summary 	Get all comics
// @Description Get all comics
// @Tags 		comics
// @Accept 		json
// @Produce 	json
// @Param 		username 	 query		string  false  "Owner of the comics"
// @Param 		title	 	 query		string  false  "Title of the comic"
// @Param 		order[price] query		string  false  "Order by price"  Enums(asc, desc)
// @Success 	200 		 {array} 	[]models.Comic
// @Failure 	503 		 {object} 	ErrorResponse{}
// @Router 		/v1/comics 	 [get]
func GetAllComics(c *fiber.Ctx) error {
	db := database.DBConn

	var comics []models.Comic
	query := db.Model(&models.Comic{})
	query.Debug()

	if user := c.Query("username"); user != "" {
		query.Where("username = ?", user)
	}

	if title := c.Query("title"); title != "" {
		query.Where("title ILIKE ?", "%"+title+"%")
	}

	if orderPrice := strings.ToLower(c.Query("order[price]")); orderPrice != "" {
		if orderPrice == "asc" {
			query.Order("price ASC")
		} else if orderPrice == "desc" {
			query.Order("price DESC")
		}
	}

	if res := query.Find(&comics); res.Error != nil {
		return c.Status(http.StatusServiceUnavailable).JSON(ErrorResponse{
			Message: res.Error.Error(),
		})
	}

	return c.JSON(comics)
}

// GetComicByID is a function to get a comic by ID
//
// @Summary 	 Get comic by ID
// @Description  Get comic by ID
// @Tags 		 comics
// @Accept 		 json
// @Produce 	 json
// @Param 		 id 			  path 	 	int  true  "Comic ID"
// @Success 	 200 			  {object}  models.Comic
// @Failure 	 404 			  {object}  ErrorResponse{}
// @Failure 	 503 			  {object}  ErrorResponse{}
// @Router 		 /v1/comics/{id}  [get]
func GetComicById(c *fiber.Ctx) error {
	id := c.Params("id")
	db := database.DBConn

	comic := new(models.Comic)
	if err := db.First(&comic, id).Error; err != nil {
		switch err.Error() {
		case "record not found":
			return c.Status(http.StatusNotFound).JSON(ErrorResponse{
				Message: fmt.Sprintf("Comic with ID %v not found", id),
			})
		default:
			return c.Status(http.StatusServiceUnavailable).JSON(ErrorResponse{
				Message: err.Error(),
			})
		}
	}

	return c.JSON(comic)
}

// AddComic adds a new comic to the users collection
//
// @Summary 	Add a new comic
// @Description Add a new comic
// @Security BearerToken
// @Tags 		comics
// @Accept 		json
// @Produce 	json
// @Param 		comic 	  	body 	  models.NewComic  true  "Add a comic"
// @Success 	201 	  	{object}  models.Comic
// @Failure 	422 	  	{object}  ErrorResponse{}
// @Router 		/v1/comics	[post]
func AddComic(c *fiber.Ctx) error {
	db := database.DBConn

	comic := new(models.Comic)
	if err := c.BodyParser(&comic); err != nil {
		return c.Status(http.StatusUnprocessableEntity).JSON(ErrorResponse{
			Message: err.Error(),
		})
	}

	user := c.Locals("user").(*jwt.Token)
	claims := user.Claims.(jwt.MapClaims)
	username := claims["username"].(string)
	comic.Username = username

	db.Create(comic)

	return c.Status(http.StatusCreated).JSON(comic)
}

// DeleteComic function removes a comic by ID
//
// @Summary 	Remove comic by ID
// @Description Remove comic by ID
// @Security BearerToken
// @Tags 		comics
// @Accept 		json
// @Produce 	json
// @Param 		id  path  int  true  "Comic ID"
// @Success 	200
// @Failure 	404 			 {object}  ErrorResponse{}
// @Failure 	503 			 {object}  ErrorResponse{}
// @Router 		/v1/comics/{id}  [delete]
func DeleteComic(c *fiber.Ctx) error {
	id := c.Params("id")
	db := database.DBConn

	user := c.Locals("user").(*jwt.Token)
	claims := user.Claims.(jwt.MapClaims)
	username := claims["username"].(string)

	comic := new(models.Comic)
	if err := db.First(&comic, id).Error; err != nil {
		switch err.Error() {
		case "record not found":
			return c.Status(http.StatusNotFound).JSON(ErrorResponse{
				Message: fmt.Sprintf("Comic with ID %v not found", id),
			})
		default:
			return c.Status(http.StatusServiceUnavailable).JSON(ErrorResponse{
				Message: err.Error(),
			})
		}
	}

	if comic.Username != username {
		return c.Status(http.StatusForbidden).JSON(ErrorResponse{
			Message: "You are not the owner of this item",
		})
	}

	db.Delete(&comic)

	return nil
}

// UpdateComic function updates a comic by ID
//
// @Summary 	Update comic by ID
// @Description Update comic by ID
// @Security BearerToken
// @Tags 		comics
// @Accept 		json
// @Produce 	json
// @Param 		id    			 path 	  int		  	  true "Comic ID"
// @Param 		comic 			 body 	  models.NewComic true "Add a comic"
// @Success 	200 	  	     {object} models.Comic
// @Failure 	404 			 {object} ErrorResponse{}
// @Failure 	503 			 {object} ErrorResponse{}
// @Router 		/v1/comics/{id}  [patch]
func UpdateComic(c *fiber.Ctx) error {
	id := c.Params("id")
	db := database.DBConn

	user := c.Locals("user").(*jwt.Token)
	claims := user.Claims.(jwt.MapClaims)
	username := claims["username"].(string)

	comic := new(models.Comic)
	if err := db.First(&comic, id).Error; err != nil {
		switch err.Error() {
		case "record not found":
			return c.Status(http.StatusNotFound).JSON(ErrorResponse{
				Message: fmt.Sprintf("Comic with ID %v not found", id),
			})
		default:
			return c.Status(http.StatusServiceUnavailable).JSON(ErrorResponse{
				Message: err.Error(),
			})
		}
	}

	updates := new(models.Comic)
	if err := c.BodyParser(&updates); err != nil {
		return c.Status(http.StatusUnprocessableEntity).JSON(ErrorResponse{
			Message: err.Error(),
		})
	}

	if comic.Username != username {
		return c.Status(http.StatusForbidden).JSON(ErrorResponse{
			Message: "You are not the owner of this item",
		})
	}

	db.Model(&comic).Updates(updates)

	return c.JSON(comic)
}

File: rs-impl/src/api/comic.rs
use chrono::NaiveDate;
use diesel::prelude::*;
use poem::{web::Data, Result};
use poem_openapi::param::{Path, Query};
use poem_openapi::{payload::Json, OpenApi};
use poem_openapi::{ApiResponse, Object};

use crate::api::{ApiTags, OrderDirection};
use crate::auth::JwtAuthorization;
use crate::database::DbPool;
use crate::models::{Comic, NewComic, ReplaceComic, UpdateComic};

const JS_SOURCE: &str = r#"
fetch("/api/comics")
    .then(res => res.json())
    .then(data => console.log(data));
"#;

pub struct ComicsApi;

#[OpenApi(tag = "ApiTags::Comics")]
impl ComicsApi {
    /// Get all comics
    ///
    /// The request must be authorized by a JWT token, but all records are freely
    /// available to view.
    #[oai(
        path = "/comics",
        method = "get",
        code_sample(lang = "js", source = "JS_SOURCE")
    )]
    async fn get_comics(
        &self,
        pool: Data<&DbPool>,
        page: Query<Option<i64>>,
        #[oai(name = "order[price]")] order_price: Query<Option<OrderDirection>>,
    ) -> Result<PaginateResponse> {
        use crate::schema::comics::dsl::*;

        let limit = 10;
        let page = page.unwrap_or(1) - 1;
        let offset = limit * page;

        let query = comics.limit(limit).offset(offset).into_boxed();
        let query = match order_price.0 {
            Some(OrderDirection::Desc) => query.order(price.desc()),
            Some(OrderDirection::Asc) => query.order(price.asc()),
            None => query,
        };

        let conn = pool.get().map_err(|err| {
            PaginateResponse::ServerError(Json(ErrorResponse {
                message: err.to_string(),
            }))
        })?;

        let res = query.load::<Comic>(&conn).map_err(|err| {
            PaginateResponse::ServerError(Json(ErrorResponse {
                message: err.to_string(),
            }))
        })?;

        Ok(PaginateResponse::Success(Json(res)))
    }

    /// Create a comic
    ///
    /// Add a comic to the currently authenticated user's collection
    #[oai(path = "/comics", method = "post")]
    async fn create_comic(
        &self,
        pool: Data<&DbPool>,
        comic_data: Json<NewComicRequest>,
        jwt: JwtAuthorization,
    ) -> Result<CreateResponse> {
        use crate::schema::comics::table;

        let conn = pool.get().map_err(|err| {
            CreateResponse::ServerError(Json(ErrorResponse {
                message: err.to_string(),
            }))
        })?;

        let res: Comic = diesel::insert_into(table)
            .values(&comic_data.0.for_user(jwt.0.username))
            .get_result(&conn)
            .map_err(|err| {
                CreateResponse::ServerError(Json(ErrorResponse {
                    message: err.to_string(),
                }))
            })?;

        Ok(CreateResponse::Created(Json(res)))
    }

    /// Get a comic by ID
    ///
    /// The request must be authorized by a JWT token, but all records are freely
    /// available to view.
    #[oai(path = "/comics/:id", method = "get")]
    async fn get_comic_by_id(&self, pool: Data<&DbPool>, id: Path<i32>) -> Result<FindResponse> {
        use crate::schema::comics::dsl::comics;

        let conn = pool.get().map_err(|err| {
            FindResponse::ServerError(Json(ErrorResponse {
                message: err.to_string(),
            }))
        })?;

        let res = comics.find(id.0).first(&conn).map_err(|err| {
            FindResponse::ServerError(Json(ErrorResponse {
                message: err.to_string(),
            }))
        })?;

        Ok(FindResponse::Success(Json(res)))
    }

    /// Replace a comic
    ///
    /// All fields are required. Must be the owner of the comic to be able to successfully update.
    #[oai(path = "/comics/:id", method = "put")]
    async fn replace_comic(
        &self,
        pool: Data<&DbPool>,
        id: Path<i32>,
        comic_data: Json<ReplaceComic>,
        jwt: JwtAuthorization,
    ) -> Result<ReplaceResponse> {
        use crate::schema::comics::dsl::comics;

        let conn = pool.get().map_err(|err| {
            ReplaceResponse::ServerError(Json(ErrorResponse {
                message: err.to_string(),
            }))
        })?;

        let comic: Comic = comics.find(id.0).first(&conn).map_err(|err| {
            ReplaceResponse::ServerError(Json(ErrorResponse {
                message: err.to_string(),
            }))
        })?;

        if comic.username != jwt.0.username {
            return Err(ReplaceResponse::Forbidden(Json(ErrorResponse {
                message: "You do not own this item".into(),
            })))?;
        }

        let res: Comic = diesel::update(comics.find(comic.id))
            .set(&comic_data.0)
            .get_result(&conn)
            .map_err(|err| {
                ReplaceResponse::ServerError(Json(ErrorResponse {
                    message: err.to_string(),
                }))
            })?;

        Ok(ReplaceResponse::Success(Json(res)))
    }

    /// Update a comic
    ///
    /// All fields are optional. Must be the owner of the comic to be able to successfully update.
    #[oai(path = "/comics/:id", method = "patch")]
    async fn update_comic(
        &self,
        pool: Data<&DbPool>,
        id: Path<i32>,
        comic_data: Json<UpdateComic>,
        jwt: JwtAuthorization,
    ) -> Result<UpdateResponse> {
        use crate::schema::comics::dsl::comics;

        let conn = pool.get().map_err(|err| {
            UpdateResponse::ServerError(Json(ErrorResponse {
                message: err.to_string(),
            }))
        })?;

        let comic: Comic = comics.find(id.0).first(&conn).map_err(|err| {
            UpdateResponse::ServerError(Json(ErrorResponse {
                message: err.to_string(),
            }))
        })?;

        if comic.username != jwt.0.username {
            return Err(UpdateResponse::Forbidden(Json(ErrorResponse {
                message: "You do not own this item".into(),
            })))?;
        }

        let res: Comic = diesel::update(comics.find(comic.id))
            .set(&comic_data.0)
            .get_result(&conn)
            .map_err(|err| {
                UpdateResponse::ServerError(Json(ErrorResponse {
                    message: err.to_string(),
                }))
            })?;

        Ok(UpdateResponse::Success(Json(res)))
    }

    /// Delete a comic
    ///
    /// Must be the owner of the comic to be able to successfully delete.
    #[oai(path = "/comics/:id", method = "delete")]
    async fn delete_comic(
        &self,
        pool: Data<&DbPool>,
        id: Path<i32>,
        jwt: JwtAuthorization,
    ) -> Result<DeleteResponse> {
        use crate::schema::comics::dsl::comics;

        let conn = pool.get().map_err(|err| {
            DeleteResponse::ServerError(Json(ErrorResponse {
                message: err.to_string(),
            }))
        })?;

        let comic: Comic = comics.find(id.0).first(&conn).map_err(|err| {
            DeleteResponse::ServerError(Json(ErrorResponse {
                message: err.to_string(),
            }))
        })?;

        if comic.username != jwt.0.username {
            return Err(DeleteResponse::Forbidden(Json(ErrorResponse {
                message: "You do not own this item".into(),
            })))?;
        }

        diesel::delete(comics.find(comic.id))
            .execute(&conn)
            .map_err(|err| {
                DeleteResponse::ServerError(Json(ErrorResponse {
                    message: err.to_string(),
                }))
            })?;

        Ok(DeleteResponse::Success)
    }
}

#[derive(Object)]
struct NewComicRequest {
    pub username: String,
    pub title: String,
    pub issue_number: String,
    pub main_character: String,
    pub genre: String,
    pub cover_year: NaiveDate,
    pub publisher: String,
    pub grade: f64,
    pub price: f64,
    pub image_url: String,
}

impl NewComicRequest {
    fn for_user(self, username: String) -> NewComic {
        NewComic {
            username,
            title: self.title,
            issue_number: self.issue_number,
            main_character: self.main_character,
            genre: self.genre,
            cover_year: self.cover_year,
            publisher: self.publisher,
            grade: self.grade,
            price: self.price,
            image_url: self.image_url,
        }
    }
}

#[derive(Debug, Object)]
struct ErrorResponse {
    message: String,
}

#[derive(ApiResponse)]
enum PaginateResponse {
    #[oai(status = 200)]
    Success(Json<Vec<Comic>>),
    #[oai(status = 401)]
    _Unauthorized(Json<ErrorResponse>),
    #[oai(status = 500)]
    ServerError(Json<ErrorResponse>),
}

#[derive(ApiResponse)]
enum CreateResponse {
    #[oai(status = 201)]
    Created(Json<Comic>),
    #[oai(status = 401)]
    _Unauthorized(Json<ErrorResponse>),
    #[oai(status = 500)]
    ServerError(Json<ErrorResponse>),
}

#[derive(ApiResponse)]
enum FindResponse {
    #[oai(status = 200)]
    Success(Json<Comic>),
    #[oai(status = 401)]
    _Unauthorized(Json<ErrorResponse>),
    #[oai(status = 500)]
    ServerError(Json<ErrorResponse>),
}

#[derive(ApiResponse)]
enum ReplaceResponse {
    #[oai(status = 200)]
    Success(Json<Comic>),
    #[oai(status = 401)]
    _Unauthorized(Json<ErrorResponse>),
    #[oai(status = 403)]
    Forbidden(Json<ErrorResponse>),
    #[oai(status = 500)]
    ServerError(Json<ErrorResponse>),
}

#[derive(ApiResponse)]
enum UpdateResponse {
    #[oai(status = 200)]
    Success(Json<Comic>),
    #[oai(status = 401)]
    _Unauthorized(Json<ErrorResponse>),
    #[oai(status = 403)]
    Forbidden(Json<ErrorResponse>),
    #[oai(status = 500)]
    ServerError(Json<ErrorResponse>),
}

#[derive(ApiResponse)]
enum DeleteResponse {
    #[oai(status = 204)]
    Success,
    #[oai(status = 401)]
    _Unauthorized(Json<ErrorResponse>),
    #[oai(status = 403)]
    Forbidden(Json<ErrorResponse>),
    #[oai(status = 500)]
    ServerError(Json<ErrorResponse>),
}
